client/tracing/src/lib.rs

```diff
@@ -0,0 +1,194 @@
+	/// either with a level "paint=trace"
+	/// or without "paint".
+	pub fn new(receiver: TracingReceiver, targets: &String) -> Self {
+		let targets: Vec<_> = targets.split(',').map(|s| parse_target(s)).collect();
+		ProfilingSubscriber {
+			next_id: AtomicU64::new(1),
+			targets,
+			receiver,
+			span_data: Mutex::new(HashMap::new()),
+		}
+	}
+}
+
+// Default to TRACE if no level given or unable to parse Level
+// We do not support a global `Level` currently
+fn parse_target(s: &str) -> (String, Level) {
+	match s.find("=") {
+		Some(i) => {
+			let target = s[0..i].to_string();
+			if s.len() > i {
+				let level = s[i + 1..s.len()].parse::<Level>().unwrap_or(Level::TRACE);
+				(target, level)
+			} else {
+				(target, Level::TRACE)
+			}
+		}
+		None => (s.to_string(), Level::TRACE)
+	}
+}
+
+impl Subscriber for ProfilingSubscriber {
+	fn enabled(&self, metadata: &Metadata<'_>) -> bool {
+		for t in &self.targets {
+			if metadata.target().starts_with(t.0.as_str()) && metadata.level() <= &t.1 {
+				log::debug!("Enabled target: {}, level: {}", metadata.target(), metadata.level());
+				return true;
+			} else {
+				log::debug!("Disabled target: {}, level: {}", metadata.target(), metadata.level());
+			}
+		}
+		false
+	}
+
+	fn new_span(&self, attrs: &Attributes<'_>) -> Id {
+		let id = self.next_id.fetch_add(1, Ordering::Relaxed);
+		let span_datum = SpanDatum {
+			id: id,
+			name: attrs.metadata().name(),
+			target: attrs.metadata().target().to_string(),
+			level: attrs.metadata().level().clone(),
+			line: attrs.metadata().line().unwrap_or(0),
+			start_time: Instant::now(),
+			overall_time: Duration::from_nanos(0),
+		};
+		self.span_data.lock().insert(id, span_datum);
+		Id::from_u64(id)
+	}
+
+	fn record(&self, _span: &Id, _values: &Record<'_>) {}
+
+	fn record_follows_from(&self, _span: &Id, _follows: &Id) {}
+
+	fn event(&self, _event: &Event<'_>) {}
+
+	fn enter(&self, span: &Id) {
+		let mut span_data = self.span_data.lock();
+		let start_time = Instant::now();
+		if let Some(mut s) = span_data.get_mut(&span.into_u64()) {
+			s.start_time = start_time;
+		} else {
+			log::warn!("Tried to enter span {:?} that has already been closed!", span);
+		}
+	}
+
+	fn exit(&self, span: &Id) {
+		let mut span_data = self.span_data.lock();
+		let end_time = Instant::now();
+		if let Some(mut s) = span_data.get_mut(&span.into_u64()) {
+			s.overall_time = end_time - s.start_time + s.overall_time;
+		}
+	}
+
+	fn try_close(&self, span: Id) -> bool {
+		let mut span_data = self.span_data.lock();
+		if let Some(data) = span_data.remove(&span.into_u64()) {
+			self.send_span(data);
+		};
+		true
+	}
+}
+
+impl ProfilingSubscriber {
+	fn send_span(&self, span_datum: SpanDatum) {
+		match self.receiver {
+			TracingReceiver::Log => print_log(span_datum),
+			TracingReceiver::Telemetry => send_telemetry(span_datum),
+		}
+	}
+}
+
+fn print_log(span_datum: SpanDatum) {
+	let message = format!(
+		"Tracing: {} {}: {}, line: {}, time: {}ns",
```

Comment:
`{} ns` looks better and can facilitate parsing. Maybe try using tabs so times/columns can be easily compared.

Reply:
I don't think tabs will work well because there can be a huge variation in module path

Reply:
but have added your suggestion for spacing the units